server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/UsersRepositoryAliasingForwarding.java

```diff
@@ -71,43 +85,68 @@ public void setUsersRepository(UsersRepository usersRepository) {
     }
 
-    /**
-     * Return null when the mail should be GHOSTed, the username string when it
-     * should be changed due to the ignoreUser configuration.
-     * 
-     * @param sender
-     * @param recipient
-     * @param message
-     * @throws MessagingException
-     */
-    public Collection<MailAddress> processMail(MailAddress sender, MailAddress recipient, MimeMessage message) throws MessagingException {
-        if (recipient == null) {
-            throw new IllegalArgumentException("Recipient for mail to be spooled cannot be null.");
+    @Inject
+    public final void setProcessor(RecipientRewriteTableProcessor processor) {
+        this.processor = processor;
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        Preconditions.checkNotNull(mail);
+        Preconditions.checkNotNull(mail.getMessage());
+        MimeMessage message = mail.getMessage();
+
+        Builder<MailAddress> errors = ImmutableList.builder();
+        processor.replaceReturnPath(mail);
+
+        Function<MailAddress, Collection<MailAddress>> processMail = getProcessMailFunction(mail, errors);
+        
+        ImmutableList<MailAddress> newRecipients = processor.rewriteRecipients(mail, processMail);
+
+        if (!errors.build().isEmpty()) {
+            getMailetContext().sendMail(mail.getSender(), errors.build(), message, Mail.ERROR);
         }
-        if (message == null) {
-            throw new IllegalArgumentException("Mail message to be spooled cannot be null.");
+
+        if (newRecipients.isEmpty()) {
+            mail.setState(Mail.GHOST);
         }
 
-        if (usersRepository instanceof RecipientRewriteTable) {
-            Mappings mappings;
-            try {
-                mappings = ((RecipientRewriteTable) usersRepository).getMappings(recipient.getLocalPart(), recipient.getDomain());
-            } catch (ErrorMappingException e) {
-                String errorBuffer = "A problem as occoured trying to alias and forward user " + recipient + ": " + e.getMessage();
-                throw new MessagingException(errorBuffer);
-            } catch (RecipientRewriteTableException e) {
-                String errorBuffer = "A problem as occoured trying to alias and forward user " + recipient + ": " + e.getMessage();
-                throw new MessagingException(errorBuffer);
-            }
+        mail.setRecipients(newRecipients);
+
+    }
+
+    private Function<MailAddress, Collection<MailAddress>> getProcessMailFunction(final Mail mail, final Builder<MailAddress> errors) {
+        return new Function<MailAddress, Collection<MailAddress>>() {
+            @Override
+            public Collection<MailAddress> apply(MailAddress recipient) {
+                if (recipient == null) {
+                    throw new IllegalArgumentException("Recipient for mail to be spooled cannot be null.");
+                }
+
+                if (usersRepository instanceof RecipientRewriteTable) {
```

Comment:
There are so many concrete of the usersRepository and it means just care if it's RRT. Any way, the UsersRepositoryAliasingForwarding.java is deprecated and it had comment at the top
>@deprecated use org.apache.james.transport.mailets.RecipientRewriteTable

So later on, there is only one RRTMailet